internal/client/main.go

```diff
@@ -35,10 +37,89 @@ func run(cfg *config) error {
 	}
 	defer conn.Close()
 	client := echo.NewEchoServiceClient(conn)
-	resp, err := client.Hello(context.Background(), &echo.HelloRequest{Message: cfg.Message})
+	switch cfg.Stream {
+	case "none":
+		return runUnary(client, cfg)
+	case "client":
+		return runClientStream(client, cfg)
+	case "server":
+		return runServerStream(client, cfg)
+	case "bidi":
+		return runBiDiStream(client, cfg)
+	}
+	return nil
+}
+
+func runUnary(client echo.EchoServiceClient, cfg *config) error {
+	if len(cfg.Messages) > 1 {
+		return errors.New("Only one message allowed for unary client requests")
+	}
+	resp, err := client.Hello(context.Background(), &echo.HelloRequest{Message: cfg.Messages[0]})
+	if err != nil {
+		return err
+	}
+	_, err = fmt.Fprintf(cfg.out, "Response: %s\n", resp.Response)
+	return err
+}
+
+func runClientStream(client echo.EchoServiceClient, cfg *config) error {
+	stream, err := client.HelloClientStream(context.Background())
+	if err != nil {
+		return err
+	}
+	for _, msg := range cfg.Messages {
+		if err := stream.Send(&echo.HelloRequest{Message: msg}); err != nil {
+			return err
+		}
+	}
+	resp, err := stream.CloseAndRecv()
 	if err != nil {
 		return err
 	}
 	_, err = fmt.Fprintf(cfg.out, "Response: %s\n", resp.Response)
 	return err
 }
+
+func runServerStream(client echo.EchoServiceClient, cfg *config) error {
+	if len(cfg.Messages) > 1 {
+		return errors.New("Only one message allowed for unary client requests")
+	}
+	req := &echo.HelloRequest{Message: cfg.Messages[0]}
+	stream, err := client.HelloServerStream(context.Background(), req)
+	if err != nil {
+		return err
+	}
+	for {
+		resp, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		_, err = fmt.Fprintf(cfg.out, "Response: %s\n", resp.Response)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func runBiDiStream(client echo.EchoServiceClient, cfg *config) error {
+	stream, err := client.HelloBiDiStream(context.Background())
+	if err != nil {
+		return err
+	}
+	for _, msg := range cfg.Messages {
+		if err := stream.Send(&echo.HelloRequest{Message: msg}); err != nil {
+			return err
+		}
+		resp, err := stream.Recv()
```

Comment:
is this receive blocking? i thought it should be possibly to send a bunch of messages from client and only get a response every say X messages?
(as in receive send and receive should run over their own channels / in their own go routines or so?)

Reply:
a response every X messages is not yet possible in pony. I would need to implement that `batch: true` protocol first.

and yes, this receive is blocking. whether these need to be separate goroutines or not would depend on the bidi method and how it works. For the echo service, it accepts one message from the client and returns one response to the client. So on the client, this send then receive then loop logic works fine.

if the method were one where one request could cause an indeterminate number of responses, or if the requests and responses were completely decoupled, then you may well need separate goroutines. but that is not the case for the echo service, and this is a client for the echo service.

I did start writing it as a goroutine, but it is just complicated for no good reason here. perhaps once i have batching implemented, I may build a fancier service where sending/receiving are decoupled and need to run asynchronously.

Reply:
one thing pony wont be able to do is send streaming server responses without first receiving a request. If you were implementing some sort of chat where each end can send messages to the other end independently, you'd need goroutines. but for pony, we will never receive a message unsolicited, so if you can tell when you have received the last response (assuming one request can generate many responses), you can do without the concurrency and just loop "send 1 then receive N".